### DISCUSSION TITLE
Update Markers onClick

### DISCUSSION THREAD

**Discussion by ksurell** [iTwin/itwinjs-core#2587] [Q&A]
I would like to update my Markers image when I click a Widget. The click function will make a rest call to update a device and I would like the markers icons to reflect the change. Currently I call (I only have one decorator):

       for (let i = 0; i < IModelApp.viewManager.decorators.length; ++i){
            IModelApp.viewManager.dropDecorator(IModelApp.viewManager.decorators[i]); 
        }
        IModelApp.viewManager.addDecorator(new SmartDeviceDecorator(IModelApp.viewManager.selectedView!));

This kind of work but I'm not sure this is the best approach.

/Krister

**Comment by pmconne** [comment]
Did you consider calling [setImage](https://www.itwinjs.org/reference/imodeljs-frontend/views/marker/setimage/) on the marker(s) to change the image?

**Reply by ksurell** [reply]
Of course, but I could not find a way to access the markers directly (I still suck at Javascript/React!). When I call addDecorator() everything is recreated which works but I realize it might be overkill, hence my question.

/Krister

**Comment by pmconne** [comment]
Given the name "SmartDeviceDecorator" I assume your code is based on one of our sandboxes like [this one](https://www.itwinjs.org/sandboxes/ThomasDicarlo/World%20Decoration%20Markers)?
SmartDeviceDecorator has a list (okay, two lists in this example) of Markers.
Modify your copy of it to either provide access  to those Markers from outside, or to supply a method your click handler can invoke to update their images.

**Comment by ksurell** [comment]
You'r right and I have already added an UpdateImage(state) method to this class. My problem is how to get hold of the markers from my click handler. I cannot find any reference to them from ViewManager, Decorator or Viewport.
/Krister

**Reply by pmconne** [reply]
![image](https://user-images.githubusercontent.com/22944042/138904974-8e48d1b6-6e4c-45fc-8586-c06bdcefc064.png)

If this is hard because React then I can't help - I don't use React.

**Reply by ksurell** [reply]
Great! I missed this since I was just looking at the default classes.
I also found that I could override drawDecortation() in Marker:

public drawDecoration(ctx: CanvasRenderingContext2D){
  super.drawDecoration(ctx);
  this.SetImage(hueLights[this._smartDeviceId]);
}
